router/httpTriggers.go

```diff
@@ -135,75 +142,77 @@ func (ts *HTTPTriggerSet) watchTriggers() {
-	rv := ""
-	for {
-		wi, err := ts.fissionClient.Httptriggers(metav1.NamespaceAll).Watch(metav1.ListOptions{
-			ResourceVersion: rv,
-		})
-		if err != nil {
-			log.Fatalf("Failed to watch http trigger list: %v", err)
-		}
-
-		for {
-			ev, more := <-wi.ResultChan()
-			if !more {
-				// restart watch from last rv
-				break
-			}
-			if ev.Type == watch.Error {
-				// restart watch from the start
-				rv = ""
-				time.Sleep(time.Second)
-				break
-			}
-			ht := ev.Object.(*tpr.Httptrigger)
-			rv = ht.Metadata.ResourceVersion
-			ts.syncTriggers()
-		}
+	watchlist := k8sCache.NewListWatchFromClient(ts.tprClient, "httptriggers", metav1.NamespaceDefault, fields.Everything())
+	listWatch := &k8sCache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return watchlist.List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return watchlist.Watch(options)
+		},
 	}
+	resyncPeriod := 30 * time.Second
+	_, controller := k8sCache.NewInformer(listWatch, &tpr.Httptrigger{}, resyncPeriod,
+		k8sCache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				ts.syncTriggers()
+			},
+			DeleteFunc: func(obj interface{}) {
+				ts.syncTriggers()
+			},
+			UpdateFunc: func(oldObj interface{}, newObj interface{}) {
+				ts.syncTriggers()
+			},
+		})
+	stop := make(chan struct{})
+	defer func() {
+		stop <- struct{}{}
+	}()
+	controller.Run(stop)
 }
 
 func (ts *HTTPTriggerSet) watchFunctions() {
-	rv := ""
-	for {
-		wi, err := ts.fissionClient.Functions(metav1.NamespaceAll).Watch(metav1.ListOptions{
-			ResourceVersion: rv,
-		})
-		if err != nil {
-			log.Fatalf("Failed to watch function list: %v", err)
-		}
+	ts.syncTriggers()
 
-		for {
-			ev, more := <-wi.ResultChan()
-			if !more {
-				// restart watch from last rv
-				break
-			}
-			if ev.Type == watch.Error {
-				// restart watch from the start
-				rv = ""
-				time.Sleep(time.Second)
-				break
-			}
-			fn := ev.Object.(*tpr.Function)
-			rv = fn.Metadata.ResourceVersion
-
-			// update resolver function reference cache
-			for key, rr := range ts.resolver.copy() {
-				if key.functionReference.Name == fn.Metadata.Name &&
-					rr.functionMetadata.ResourceVersion != fn.Metadata.ResourceVersion {
-					err := ts.resolver.delete(key.namespace, &key.functionReference)
-					if err != nil {
-						log.Printf("Error deleting functionReferenceResolver cache: %v", err)
+	watchlist := k8sCache.NewListWatchFromClient(ts.tprClient, "functions", metav1.NamespaceDefault, fields.Everything())
+	listWatch := &k8sCache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return watchlist.List(options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return watchlist.Watch(options)
+		},
+	}
+	resyncPeriod := 30 * time.Second
+	_, controller := k8sCache.NewInformer(listWatch, &tpr.Function{}, resyncPeriod,
```

Comment:
How about keeping the store -- and keeping it in  a new `ts.store`? And then using it in `syncTriggers`?